File: kotlin/src/main/kotlin/io/digitalstream/advanced/Generics.kt
/**
 * Created by sebby on 9/25/16.
 */
package io.digitalstream.advanced

object Generics {

  class Point<T>(val x: T, val y: T) {
    fun swap(): Point<T> {
      return Point(y, x)
    }
  }

  fun run(): Unit {
    val p = Point(0.0, 1.0).swap()
    println("point ${p.x} ${p.y}")
  }
}

File: scala/src/main/scala/io/digitalstream/advanced/Generics.scala
/**
  * Created by sebby on 9/25/16.
  */
package io.digitalstream.advanced

object Generics {

  class Point[T](val x: T, val y: T) {
    def swap(): Point[T] = {
      new Point(y, x)
    }
  }

  def run(): Unit = {
    val p = new Point(0.0, 1.0).swap()
    println(s"point ${p.x} ${p.y}")
  }
}
